test/common_distributed.py

```diff
@@ -121,39 +123,66 @@ def wrapper(self):
     @classmethod
     def setUpClass(cls):
         for attr in dir(cls):
-            if attr.startswith('test'):
+            if attr.startswith("test"):
                 fn = getattr(cls, attr)
                 setattr(cls, attr, cls.join_or_run(fn))
 
     def setUp(self):
         super(MultiProcessTestCase, self).setUp()
         self.skip_return_code_checks = []
         self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
-        self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
+        self.file_name = tempfile.NamedTemporaryFile(delete=False).name
+        self._spawn_process()
 
     def tearDown(self):
         super(MultiProcessTestCase, self).tearDown()
         for p in self.processes:
             p.terminate()
 
-    def _spawn_process(self, rank):
-        name = 'process ' + str(rank)
-        process = multiprocessing.Process(target=self._run, name=name, args=(rank,))
-        process.start()
-        return process
+    def _current_test_name(self):
+        # self.id() == e.g. '__main__.TestDistributed.TestAdditive.test_get_rank'
+        return self.id().split(".")[-1]
+
+    def _spawn_process(self):
+        test_name = self._current_test_name()
+        # torch.multiprocessing.spawn does not support python2
+        if sys.version_info >= (3, 0):
+            self.spawncontext = torch.multiprocessing.spawn(
+                self.__class__._run,
+                (test_name, self.file_name),
+                nprocs=int(self.world_size),
+                join=False,
+            )
+            self.processes = self.spawncontext.processes
```

Comment:
I don't see this calling `spawncontext.join()` anywhere, which is the API contract. If this doesn't use `torch.multiprocessing.spawn` as intended, I don't see the added value of using it in the first place.

What did this solve over the previous approach with just `multiprocessing.set_start_method("spawn")`?

Reply:
spawncontext.join() failed to join in my previous tests because multiprocessing.connection.wait() failed for some processes, let me try one more time to use it

torch.multiprocessing.spawn resolved two weird test failures in multiprocessing.set_start_method("spawn"):  1) asan test, although I skipped asan 2) Numpy + Intel MKL incompatibility error